### ISSUE TITLE
EI_EXPOSE_REP and EI_EXPOSE_REP2 occur if using Guava Immutable objects

### ISSUE THREAD

**Issue by callum-kilby** [spotbugs/spotbugs#1601]
We are using [Guava Immutable objects](https://guava.dev/releases/snapshot/api/docs/com/google/common/collect/ImmutableCollection.html) in our project, and from SpotBugs 4.3.0 we have SpotBugs failures EI_EXPOSE_REP and EI_EXPOSE_REP2 when directly setting/retrieving class fields as Immutable objects.

For example:

```java
private Map<String, String> overrides;

public CLISetup(String commandLine, ImmutableMap<String, String> overrides) {
    this.commandLine = commandLine;
    this.overrides = overrides;
}
```
results in an EI_EXPOSE_REP2 failure from this.overrides = overrides.

This failure is incorrect because the object being passed in and set for the class field is shallowly immutable.

**Comment by vladimirfx** [comment]
After upgrade to 4.3.0, I have a huge amount of EI_EXPOSE_REP and EI_EXPOSE_REP2 on previously clean code. (i.e. reports on getters returning "java.time.Instant")

I was forced to disable both rules. It is possible to restore previous logic in form of a separate rule OR introduce new logic as separate rules? It will be very useful for obvious cases when an array or something like "java.util.Date" is returned or stored.

**Comment by xzel23** [comment]
I also ran into the issue with returning java.time.Instant. I don't know how this is supposed to work: how would SpotBugs know that java.time.Instant is immutable? Does it have a whitelist of classes, or is this done by analysing the bytecode? If it's the first, the immutable standard classes should be added to the list, if second, something's wrong with the byte code analyser.

**Comment by xzel23** [comment]
OK, I just saw it. `java.time.Instant` should be added to `edu.umd.cs.findbugs.util.MutableClasses.KNOWN_IMMUTABLE_CLASSES`. 

Here is a list of the classes that are immutable in that package (stated directly in the classes' Javadoc):

java.time.Duration
java.time.Instant
java.time.LocalDate
java.time.LocalDateTime
java.time.LocalTime
java.time.MonthDay
java.time.OffsetDateTime
java.time.OffsetTime
java.time.Year
java.time.YearMonth
java.time.ZonedDateTime
java.time.ZoneOffset
java.time.ZoneRegion

BUT IMHO there's a much better way (at least for the JDK classes): if SpotBugs is running on a recent JDK (= 16+) , it can check for the presence of the annotation `jdk.internal.ValueBased`. Alternatively, I can provide you with a list of all 44 so annotated classes that should be included in the KNOWN_IMMUTABLE_CLASSES list, so that it also works with older JDK versions. (To generate such a list open the annotation in IntelliJ IDEA and run a "Find Usages" search / probably also possible in Eclipse, but I don't have it here.)

And a last question: I also have this warning for Java.lang.Class. I think taht could be added too.

**Comment by xzel23** [comment]
PS: I could probably find the time to checkout the SpotBugs source and provide a patch later this week, just tell me which approach to take (and if I should open another issue "add JDK immutable classes" to KNOWN_IMMUTABLE:CLASSES).

**Comment by vladimirfx** [comment]
There are approaches to detect mutable state exposition whitelisting and blacklisting. I think before 4.3.0 detector uses blacklisting but starting from 4.3.0 it uses whitelisting. 
IMO it nearly impossible to cover all immutable classes in a white list (ex. we use extensively Kotlin immutable data classes). 

So I asked to restore previous behavior in any form - it is been very useful for my team.

**Comment by baloghadamsoftware** [comment]
I added the whole java.time package to the list of immutable packages (packages which only contain immutable classes).

**Comment by MPoorter** [comment]
@baloghadamsoftware The same happens for BigDecimal and BigInteger. Could those be added to the list as well? Unfortunately `java.math` does seem to have mutable classes in it as well.

**Comment by baloghadamsoftware** [comment]
> 
> 
> @baloghadamsoftware The same happens for BigDecimal and BigInteger. Could those be added to the list as well? Unfortunately `java.math` does seem to have mutable classes in it as well.

This is strange, because these classes do not have setter like methods. (They have methods beginning with "add" or "set", but all of them returns another instance of the same object, thus they are not considered as setters.) Anyway, I will put the whole package explicitly to the list of known packages without mutables. But first I should check why are they recognized as mutables. Maybe there is a bug?

**Comment by baloghadamsoftware** [comment]
Now I added java.math and fixed the heuristics as well. Previously I checked the return value of setter-like methods and decided that they are no setters only if they returned the type as their class. There was a test for it which passed, but it seems now that it did not work fully correctly. Now I updated this logic, if a method which has a setter-like name but does return a class instance then I consider it as a non-setter. This reduces the amount of false positives by a lot.

**Comment by baloghadamsoftware** [comment]
> 
> 
> After upgrade to 4.3.0, I have a huge amount of EI_EXPOSE_REP and EI_EXPOSE_REP2 on previously clean code. (i.e. reports on getters returning "java.time.Instant")
> 
> I was forced to disable both rules. It is possible to restore previous logic in form of a separate rule OR introduce new logic as separate rules? It will be very useful for obvious cases when an array or something like "java.util.Date" is returned or stored.

Actually, the previous logic was just a placeholder. There was a list of 4 (yes, four) classes which were considered as mutable and everything else as immutable.

**Comment by JackOfMostTrades** [comment]
I'm also seeing this trigger for uses of X509Certificate which I believe is immutable. For concreteness, getting FPs on this class:

```
import java.security.cert.X509Certificate;
public class Issue1601 {
    private X509Certificate certificate;

    public X509Certificate getCertificate() {
        return certificate;
    }

    public void setCertificate(X509Certificate certificate) {
        this.certificate = certificate;
    }
}
```

**Comment by maisonobe** [comment]
Is it possible to have a way to configure a set of our own classes that should be considered safe to store/retrieve?
I have just upgrade a big project and am completely overwhelmed with false positive.

Several classes that are immutable (no setters, everything final) are not considered immutable by the heuristic
which is a clear false positive. There are also classes that are considered too huge to be copied,
and hence accepted to be shared (think a big matrix with millions of entries for example). In this case, qualifying
them as not immutable is correct, but acceptable in some context and as references to these huge data sets
are exchanged all the time by processing algorithm, we have to put filters on all functions just to say "hey, this
is a 100megabytes image, it's normal I don't copy it".

I am not speaking of java classes, but of classes from third parties libraries. Obviously, the heuristic that detects
immutable classes is failing

The change done on an existing check by changing drastically its semantics with almost no way to configure
it is really a big problem. I simply cannot put thousands of filters on a 100000+ lines library that contains something like
thousand classes.

**Comment by JackOfMostTrades** [comment]
One more FP to report on this thread. In this case spotbugs believes this interface to be mutable (which it isn't, not that spotbugs would have any way to know since the impl isn't on the classpath):

```
public class Issue1601 {
    public static interface MyIface {
        void persistObject(String primaryKey, Object obj);
        void deleteObject(String primaryKey);
    }

    private final MyIface iface;

    public Issue1601(MyIface iface) {
        this.iface = iface;
    }
}
```

**Comment by maisonobe** [comment]
I have looked at the heuristic and found it is mainly based on names. Class names, and method names.
Unfortunately, I am working with mathematics, and "add" is not only something used for containers, but
also as an operation in an algebraic group... Of course, this is a special case, and of course the check for
the return value is an attempt to identify such cases. My point is that relying on name is not a robust check
to me, there are just too many corner cases and developers sometimes choose weird name for their methods
that are not consistent with the behaviour (and I do not speak of people writing code with non-english
method names).

The following method would be considered as safe by this heuristic, despite is is really not:
```
    public MyClass addToSelf(MyClass other) {
       field += other.field;
       return this;
    }
```

On the other hand, the following method would be considered unsafe, despite it really is:
```
  public class Element {
    public Couple append(Element other) {
      return new couple (this, other);
   }
 }
```
Perhaps adding this would improve the heuristic, but it is clearly not sufficient either, it just
detects some cases when developer has declared all fields as immutable. I hope it has less
false positive, but unfortunately it certainly has false negative (this is just some sketchy code,
I am not fluent with BCEL).
```
public static boolean immutableFields(JavaClass cls) {
  for (Field field : cls.getFields()) {
    if (!field.isFinal()) {
      return false;
    }
    if (field.getType() == Type.CLASS) {
      JavaClass fieldClass = Repository.lookupClass(field.getType().getSignature());
      if (isMutable(fieldClass))  {
        return false;
      }
    }
  }
  return true;
}
```

Anyway, I really think no heuristic could ever succeed, short of analyzing thoroughly the bytecode
of the class, so would really prefer that users could whitelist some classes in a global manner,
not in each and every <Match></Match> filter for thousands of methods.

**Comment by pbludov** [comment]
Same for `java.util.Locale`. This class has ho setters, please add it to the list of immutable classes.

**Comment by jzaralim** [comment]
The reason the Guava classes aren't recognized is because [this method](https://github.com/spotbugs/spotbugs/blob/4b7bb522495b87602678b24d4b71c350f7bf01ea/spotbugs/src/main/java/edu/umd/cs/findbugs/util/MutableClasses.java#L70) sees the `remove` and `set` functions and assumes that the classes are mutable. The Guava classes actually throw an `Unsupported operation` exception on these functions though. Can `looksLikeASetter` be updated to return false when the function just throws?

**Comment by baloghadamsoftware** [comment]
> 
> 
> Same for `java.util.Locale`. This class has ho setters, please add it to the list of immutable classes.

I checked `java.util.Locale` and it is indeed immutable. Now I improved the heuristics to disregard `static` methods because they are surely no setters. (Also non-public methods are disregarded now.)

**Comment by baloghadamsoftware** [comment]
> 
> 
> The reason the Guava classes aren't recognized is because [this method](https://github.com/spotbugs/spotbugs/blob/4b7bb522495b87602678b24d4b71c350f7bf01ea/spotbugs/src/main/java/edu/umd/cs/findbugs/util/MutableClasses.java#L70) sees the `remove` and `set` functions and assumes that the classes are mutable. The Guava classes actually throw an `Unsupported operation` exception on these functions though. Can `looksLikeASetter` be updated to return false when the function just throws?

Good point! Thank you! I improved the function to disregard methods which thrown `UnsupportedOperationException`.

**Comment by baloghadamsoftware** [comment]
> On the other hand, the following method would be considered unsafe, despite it really is:
> 
> ```
>   public class Element {
>     public Couple append(Element other) {
>       return new couple (this, other);
>    }
>  }
> ```

It is considered safe now because I changed the heuristics to disregard functions which return any reference, not only a reference to the same type as the owning class.

> Perhaps adding this would improve the heuristic, but it is clearly not sufficient either, it just
> detects some cases when developer has declared all fields as immutable. I hope it has less
> false positive, but unfortunately it certainly has false negative (this is just some sketchy code,
> I am not fluent with BCEL).
> 
> ```
> public static boolean immutableFields(JavaClass cls) {
>   for (Field field : cls.getFields()) {
>     if (!field.isFinal()) {
>       return false;
>     }
>     if (field.getType() == Type.CLASS) {
>       JavaClass fieldClass = Repository.lookupClass(field.getType().getSignature());
>       if (isMutable(fieldClass))  {
>         return false;
>       }
>     }
>   }
>   return true;
> }
> ```

A class which contains immutable members only is not necessarily immutable:

```
public class StringStore() {
    private String content;
    ...
    public void setContent(String content) {
        this.content = content;
    }
}
```

Although `String` is immutable, `StringStore` is not, because anyone can replace the string it stores.

> Anyway, I really think no heuristic could ever succeed, short of analyzing thoroughly the bytecode
> of the class, so would really prefer that users could whitelist some classes in a global manner,
> not in each and every filter for thousands of methods.

User-defined whitelist is a good idea, I will take a look how to implement it properly in SpotBugs.

**Comment by maisonobe** [comment]
> Although `String` is immutable, `StringStore` is not, because anyone can replace the string it stores.

This is not what I had in mind, but maybe I misunderstood what `field.isFinal()` does. I think your `StringStore` example would be detected by the first `if`: the field is not final.
What I thought was something like:

```
public class UnsafeStore1 {
  private SomeImmutableClass field1;
  public void hiddenSetter(SomeImmutableClass field1) {
    this.field1 = field1;
 }
}

public class UnsafeStore2 {
  private final SomeMutableClass field2;
  public SomeMutableClass notASetterButAllowsTransitiveChange() {
    return field2;
 }
}

public class SafeStore {
  private final SomeImmutableClass field;
}
```

The `UnsafeStore1` class would be considered mutable because `field1` is not declared as final, even despite the setter does not follows any recognizable naming convention; it is the first `if` in my proposal that detects it. The `UnsafeStore2` class would be considered mutable because `field2` despite being declared as final reference a mutable class; it is the second `if` in my proposal that detects it. The `SafeStore` class would be considered immutable because both `field` is declared as final and corresponds to a class itself considered immutable (or a primitive int, double, char...); it passes both `if` statements and falls back to the `return true`.

**Comment by baloghadamsoftware** [comment]
> This is not what I had in mind, but maybe I misunderstood what `field.isFinal()` does. I think your `StringStore` example would be detected by the first `if`: the field is not final.

Oh sorry, I was not careful enough when reading your code.

**Comment by baloghadamsoftware** [comment]
I checked lots of findings and now it seems that using the last patch I only find true positives. However, there are still many findings because in most programs this is not an issue, only in security-critical software. (Although I see no point in declaring a mutable field private and then exposing it in a getter. This is almost the same as declaring the field as public.)

**Comment by baloghadamsoftware** [comment]
This heuristics is far from perfect, because we may loose true positives, but the only way to get false positives is to have poorly named methods: if a method whose name is "setter-like" returns a basic type, but it is no setter then it considers it as a setter. Maybe I could change to only consider those as setter which do not return anything (return type void).

**Comment by maisonobe** [comment]
Anyway, a user-defined whitelist would still be useful for cases where the classes really are mutable,
but users really want to pass them around. In huge datasets context, this occurs a lot.

**Comment by baloghadamsoftware** [comment]
> 
> 
> Anyway, a user-defined whitelist would still be useful for cases where the classes really are mutable,
> but users really want to pass them around. In huge datasets context, this occurs a lot.

Of course, I will do it as I promised it.

**Comment by baloghadamsoftware** [comment]
> 
> 
> Anyway, a user-defined whitelist would still be useful for cases where the classes really are mutable,
> but users really want to pass them around. In huge datasets context, this occurs a lot.

First draft: #1619

**Comment by jglick** [comment]
I am going to revert the update to 4.3.0 for Jenkins plugins (https://github.com/jenkinsci/plugin-pom/pull/423) because it just produces tons of new warnings for things that are not that big of a deal, or could not reasonably be written any other way, like [this](https://github.com/jenkinsci/resource-disposer-plugin/blob/cccaee96860889fee86a6116395a0bce75325a35/src/main/java/org/jenkinsci/plugins/resourcedisposer/Disposable.java#L100-L112) for example. I am not convinced just tweaking the rules for detecting immutable classes will suffice; there are plenty of times where passing around a really mutable object is just how Java works. It is nice that #1580 (or whatever the change was) is catching more bugs for someone but can the priority perhaps be reconsidered, so that these are suppressed when you are using threshold=Medium and developers can focus on real mistakes?

**Comment by vladimirfx** [comment]
New logic is fairly useful (when polished) for **some** projects.
Why not implement it as separate rule?

**Comment by baloghadamsoftware** [comment]
> New logic is fairly useful (when polished) for **some** projects.
> Why not implement it as separate rule?

I am not sure that the old logic was useful for many projects. (Only 4 classes were considered as mutable, and everything else as immutable.) However, the rule itself is very special indeed, one direction (EI) is basically SEI CERT rule OBJ05-J, the other direction  (EI2) is its opposite. They are only important in a sensitive project. I think that we should changes its priority to "low".

**Comment by maisonobe** [comment]
> I am not sure that the old logic was useful for many projects

It was useful event if the only class considered mutable was arrays. In fact, I thought this check was only for arrays until 4.3.0 blew everything.

**Comment by baloghadamsoftware** [comment]
> > I am not sure that the old logic was useful for many projects
> 
> It was useful even if the only class considered mutable was arrays. In fact, I thought this check was only for arrays until 4.3.0 blew everything.

The we could change the priority to "low", except for arrays. is this OK?

**Comment by maisonobe** [comment]
I would really prefer a different check, that I would definitely turn off. Low priority is still something I check. I manage at least two project that are huge (>100000 lines), and critical (operational space flight dynamics), and they all have exactly zero warnings despite priority is set to quite low levels. But then 4.3.0 came.

If we start looking at a too fine level of security,then we just don't even use Java because it has reflection, and memory allocation, and recursion, and... The change in this check is just going too far in the direction that would correspond to things like embedded software, were they use C and formal proofs (and it costs millions).

I understand the balance is difficult to find, and the tipping point is different from project to project. But here the change was drastic. I also had to revert to an older version of spotBugs.

**Comment by iloveeclipse** [comment]
I've commented on similar detector already https://github.com/spotbugs/spotbugs/pull/1551#issuecomment-857001232

I think such detectors could be interesting only for limited number of people/projects, the 99% of rest of the world is just abused, so I would suggest instead of reducing the priority of the report, simply disable the detectors by default. 

Whoever wants security audit can enable them & be happy.

**Comment by baloghadamsoftware** [comment]
Maybe we should separate the error message to EI_EXPOSE_ARRAY_REF and EI_EXPOSE_MUTABLE_OBJECT_REF?

**Comment by maisonobe** [comment]
If that would allow a catch-all filter to silence EI_EXPOSE_MUTABLE_OBJECT_REF globally, yes it would be a good idea.

**Comment by baloghadamsoftware** [comment]
OK, so the proposed solution is (beside fixing the trivial issues) to separate the error messages for arrays and immutable classes and disable the latter by default?

**Comment by vladimirfx** [comment]
Maybe arrays + classes black list like 'java.util.Date'? Ideally, the black list should be configurable.

**Comment by baloghadamsoftware** [comment]
@KengoTODA and @ThrawnCA could you please take a look on this?

**Comment by ThrawnCA** [comment]
> @KengoTODA and @ThrawnCA could you please take a look on this?

Are you asking about the idea of splitting arrays from other potentially-mutable objects?

I can see that the detection logic for arrays could be different (simpler, since they have known properties), but is it really a different bug from the end user perspective?

**Comment by baloghadamsoftware** [comment]
> > @KengoTODA and @ThrawnCA could you please take a look on this?
> 
> Are you asking about the idea of splitting arrays from other potentially-mutable objects?
> 
> I can see that the detection logic for arrays could be different (simpler, since they have known properties), but is it really a different bug from the end user perspective?

Actually, I am asking about the acceptable solution. I created a fix (#1603) which fixes some real bugs in the algorithm to remove false positives plus it reduces the scope to reduce number of findings. However @KengoTODA claims that it does not fix anything. That is why I am asking for the soultion you accept. I can change this PR to only include the actual bugfixes, but what to do with the rest of the findings which are true positives but in an average Java program there are too many of them? One of the proposals was to have a separate error message for arrays and another one for other mutable objects so users can filter or disable the latter.

**Comment by rovarga** [comment]
I think most of of the Guava issues can be solved by supporting ErrorProne's `@Immutable` semantics. I have https://github.com/rovarga/spotbugs/tree/immutable-ep lined up as a stepping stone towards solving 1697.

**Comment by romani** [comment]
still a problem on 4.4.2.2

log of execution https://github.com/checkstyle/checkstyle/pull/10892
